spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -0,0 +1,517 @@
+        token.safeTransfer(to, amount);
+    }
+
+    //--------------------------------------------------------------------------
+    // External & Public write methods
+
+    /// @notice Recovers deployed funds and redeploys them. Reverts if there are no funds to deploy.
+    /// @dev Simply batches the `recover` and `deploy` functions.
+    function recoverAndRedeploy() external {
+        recover();
+        deploy();
+    }
+
+    /// @notice Deploys deposited funds. Reverts if there are no funds to deploy.
+    /// @dev Its safer to call `recover` before `deploy` so the full available balance can be deployed.
+    function deploy() public nonReentrant whenNotPaused {
+        TrancheData memory td = _tranche(perp.getDepositBond());
+        if (_rollover(perp, td) == 0) {
+            revert NoDeployment();
+        }
+    }
+
+    /// @notice Recovers deployed funds.
+    function recover() public nonReentrant whenNotPaused {
+        _redeemTranches();
+    }
+
+    /// @notice Deposits the provided asset from {msg.sender} into the vault and mints notes.
+    /// @param token The address of the asset to deposit.
+    /// @param amount The amount tokens to be deposited into the vault.
+    /// @return The amount of notes.
+    function deposit(IERC20Upgradeable token, uint256 amount) external nonReentrant whenNotPaused returns (uint256) {
+        // NOTE: The vault only accepts the underlying asset tokens.
+        if (token != underlying) {
+            revert UnexpectedAsset(token);
+        }
+
+        uint256 totalSupply_ = totalSupply();
+        uint256 notes = (totalSupply_ > 0) ? amount.mulDiv(totalSupply_, getTVL()) : (amount * INITIAL_RATE);
+
+        underlying.safeTransferFrom(_msgSender(), address(this), amount);
+        _syncAsset(underlying);
+
+        _mint(_msgSender(), notes);
+        return notes;
+    }
+
+    struct TokenAmount {
+        /// @notice The asset token redeemed.
+        IERC20Upgradeable token;
+        /// @notice The amount redeemed.
+        uint256 amount;
+    }
+
+    /// @notice Burns notes and sends a proportional share vault's assets back to {msg.sender}.
+    /// @param notes The amount of notes to be burnt.
+    /// @return The list of asset tokens and amounts redeemed.
+    function redeem(uint256 notes) external nonReentrant whenNotPaused returns (TokenAmount[] memory) {
+        uint256 totalNotes = totalSupply();
+        uint256 deployedCount_ = deployedCount();
+        uint256 earnedCount_ = earnedCount();
+        uint256 assetCount = 1 + deployedCount_ + earnedCount_;
+
+        TokenAmount[] memory redemptions = new TokenAmount[](assetCount);
+
+        // burn notes
+        _burn(_msgSender(), notes);
+
+        // aggregating vault assets to be redeemed
+        redemptions[0].token = underlying;
+        for (uint256 i = 0; i < deployedCount_; i++) {
+            redemptions[i + 1].token = deployedAt(i);
+        }
+        for (uint256 i = 0; i < earnedCount_; i++) {
```

Comment:
Think we really need this `for` loop, and the `earnedCount` function?

I think we already see this as not being able to configure the earned token after deployment, so I think we know it will always be 1 for this instance.

We assume there's 1 underlying in line 227, even though the deposit function allows you to specify which token to deposit. So we're already straddling two level of generalization. I'd just go with the simplest here.

Reply:
Makes sense 👍